### PR TITLE
document VMware Cloud auth provider

### PR DESCRIPTION
**Pull request by jamieklassen**
## Hey, I just made a Pull Request!

Just documentation, kinda sorely missing since #21337 was merged

#### :heavy_check_mark: Checklist

- [ ] ~A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))~ docs only
- [x] Added or updated documentation
- [ ] ~Tests for new functionality and regression tests for bug fixes~ docs only
- [ ] ~Screenshots attached (for UI changes)~
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by jamieklassen** [comment]
The use of this provider feels complicated compared to the others, largely because:

* it has no default auth provider factory in `@backstage/plugin-auth-backend`
* it has no associated utility API in `@backstage/core-plugin-api`

@Rugvip @tudi2d do you think it makes sense to address these two points? I can open a separate issue for this but I think the trickiness these instructions makes a nice illustration.

File: docs/auth/vmware-cloud/provider.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+```ts title="packages/app/src/apis.ts"
+import {
+  ScmIntegrationsApi,
+  scmIntegrationsApiRef,
+  ScmAuth,
+} from '@backstage/integration-react';
+import {
+  costInsightsApiRef,
+  ExampleCostInsightsClient,
+} from '@backstage/plugin-cost-insights';
```
Comment (Rugvip) [inline docs/auth/vmware-cloud/provider.md:128]: Fair amount of this example isn't related to what's being done, best trim this down a bit

File: docs/auth/vmware-cloud/provider.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  fetchApiRef,
+  githubAuthApiRef,
+  /* highlight-add-start */
+  oauthRequestApiRef,
+  /* highlight-add-end */
+} from '@backstage/core-plugin-api';
+import { AuthProxyDiscoveryApi } from './AuthProxyDiscoveryApi';
+
+/* highlight-add-start */
+export const vmwareCloudAuthApiRef: ApiRef<
```
Comment (Rugvip) [inline docs/auth/vmware-cloud/provider.md:161]: Is this provider only useful for sign-in or is  there some OAuth access delegation going on as well that would be useful to provide to plugins? Asking mainly for future work

Either way let's go ahead and add both this API ref and the factory the to the core APIs. `core-plugin-api` + `frontend-plugin-api` + `app-defaults`

Long term I'm looking to get rid of the need for API Refs for sign-in, and possibly consolidating all OAuth providers into a single API. But can't promise that's gonna be done soon enough to close this gap

File: docs/auth/vmware-cloud/provider.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    discovery: env.discovery,
+    tokenManager: env.tokenManager,
+    providerFactories: {
+      ...defaultAuthProviderFactories,
+      /* highlight-add-start */
+      vmwareCloudServices: createOAuthProviderFactory({
+        authenticator: vmwareCloudAuthenticator,
+        signInResolver:
+          commonSignInResolvers.emailLocalPartMatchingUserEntityName(),
+      }),
```
Comment (Rugvip) [inline docs/auth/vmware-cloud/provider.md:103]: We can add the provider as one of the built-in ones for the old system if you want, but tbh I don't think it'd be too useful at this point? Depends a bit on if it's used for access delegation too

File: docs/auth/vmware-cloud/provider.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+can enable the VMware Cloud provider with a small modification like:
+
+```ts title="packages/backend-next/src/index.ts"
+import { createBackend } from '@backstage/backend-defaults';
+
+const backend = createBackend();
+backend.add(import('@backstage/plugin-auth-backend'));
+/* highlight-add-start */
+backend.add(
+  import('@backstage/plugin-auth-backend-module-vmware-cloud-provider'),
```
Comment (Rugvip) [inline docs/auth/vmware-cloud/provider.md:53]: Let's keep this as is for the new system. We're going to move all modules to be separate and not be available as built-in providers of the `auth-backend` at all, so if we added this as a built-in provider in the new system we'd just force another migration in the future.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice, thank you! 👍 

I think there's an opportunity to trim away some code from the examples, but otherwise good to go.

Regarding adding more of this to core APIs - yep - see separate comments for that

File: docs/auth/vmware-cloud/provider.md (hunk 2 above)
Comment (jamieklassen) [reply docs/auth/vmware-cloud/provider.md:161]: There is a VMware-specific use case that will require access tokens for use with an external service, so "yes" to access delegation. I'll file an issue/PR for adding this utility API to the core.

File: docs/auth/vmware-cloud/provider.md (hunk 4 above)
Comment (jamieklassen) [reply docs/auth/vmware-cloud/provider.md:53]: Agreed!

File: docs/auth/vmware-cloud/provider.md (hunk 3 above)
Comment (jamieklassen) [reply docs/auth/vmware-cloud/provider.md:103]: I'm cool without adding the factory to the core, not a big deal.

File: docs/auth/vmware-cloud/provider.md (hunk 1 above)
Comment (jamieklassen) [reply docs/auth/vmware-cloud/provider.md]: Removed in 9185966

**Pull request merged**